README
================================================================================
To compile UCGretina

  $ source <G4INSTALL>/share/Geant4-9.6.1/geant4make/geant4make.sh
  $ make

Here, the file env.sh comes from your local installation of geant4. 

To use the liquid hydrogen target:

  $ make LHTARGET=1

produces the binary UCGretina_LH

================================================================================
To simulate the response of the array to a gamma ray

  Run a macro file like "test.mac" describing the measurement.

  $ UCGretina test.mac > test.log

  The last part --  "> test.log"  -- is optional, but it's a good
  idea. It saves the text messages from UCGretina in a log file.

UCGretina reads the description of the beam, reaction and gamma ray
from the macro file ("test.mac", e.g.). It writes the detector number,
energy deposited, and hit position of each gamma-ray interaction with
an active detector volume in ascii format to an output file (name
specified in the macro file).

The output file test.out contains gamma-ray tracking information for
each event that deposited energy in GRETINA. For each event there is a
header line beginning with '$':

   $ nHits Egamma Event#

where nHits is the number of hits, Egamma is the total energy of the
gamma ray, and Event# is the event number. The header line is followed
by nHits lines describing the hits with the format:

   detNum Edep x y z

where Edep is the energy (in keV) deposited in crystal detNum
by the hit. The hit positions (x, y, z) are in mm. 

================================================================================
Selected Commands

Optional commands for setting target/source parameters:

  /Target/Material
  /Target/X_length 
  /Target/Y_length
  /Target/SetPosition_Z
  /Target/Thickness
  /Target/ScaleDensity
  /Target/sourceFrame
  /Target/Sled 

Mandatory command for building a target/source frame:

  /Target/Construct

Optional commands for setting beam-tube parameters:

  /BeamTube/R_min
  /BeamTube/R_max
  /BeamTube/Length

Mandatory command for building the beam tube:

  /BeamTube/Construct

Optional commands for setting GRETINA parameters:

  /Gretina/detector/enableCapsules
  /Gretina/detector/enableCryostats
  /Gretina/Shell full/north/south

Mandatory command after setting any GRETINA parameters:

  /Gretina/update

To produce output ready for conversion to Mode 2 data:

  /Mode2/crmatFile <filename>

    The file "crmat.txt" is provided with this tarball. It specifies
    the rotation matrices and translation vectors from each crystal
    frame into the world coordinate system. UCGretina inverts this
    transformation to produce data in the crystal frames, as expected
    in Mode 2 data. 

  /Mode2/GretinaCoords

    When this command is present, the world coordinates are 
    rotated Pi/2 about z to match the standard GRETINA coordinate
    system (x = down, z = beam). This is the coordinate system
    expected in Mode 2 data.

    Interaction points are expressed in the Geant4 coordinate system
    (y = up, z = beam) by default.

  /Mode2/PositionRes <float> <unit>

    The PositionRes parameter determines the closest spacing of
    gamma-ray interaction points within each event that GRETINA can
    resolve. Specify a zero position resolution to turn off
    consolidation.

================================================================================
To visualize the array

  Run the macro file "vis.mac" an interactive session:

  $ UCGretina

  Idle> /control/execute vis/vis.mac
  Idle> exit

This generates a VRML 2 file named g4_XX.wrl which can be viewed with
a VRML viewer (like mayavi2).

================================================================================

To simulate, sort, and fit in-beam gamma-ray spectra, see the fit and
fitLH subdirectories.

================================================================================

To simulate calibration source measurements and compare them with
measurements, see the eu152 subdirectory.

================================================================================

Changelog
================================================================================
Version 2.1
================================================================================

Added macro file commands to set LH target parameters: cell type, window
	bulge thickness, and tilt angle.

Added thin LH target (50 mg/cm^2 nominal) and empty target cell body.

Added the "white" source type (uniform energy distribution between user-set
	limits; default: 100 keV ... 10 MeV).

TrackingAction now writes dta in % (for GrROOT compatibility).

Fine-tuning of examples.

Minor fixes to macro file commands (and documentation) for the incoming
	and outgoing beams.

Target is now invisible if made of G4_Galactic.

================================================================================
Version 2.0
================================================================================
L. Riley April 2013

Mode 2 output with decomposed gamma-ray events (GEB type 1), simulated
    S800 events (GEB type 9), and events with energies, emission
    positions, and emission directions of simulated gammas (GEB type
    11). Added TrackingAction and EventInformation classes to store
    outgoing reaction product and emitted gamma-ray information for output
    by the EventAction class.

Now reads the binary crmat file (crmat.LINUX).

Added G4Timers to the main and EventAction to report on overall timing and
	progress.

Updated examples using GrROOT package for sorting.

================================================================================
Version 1.3.0
================================================================================
L. Riley  March 2013

New solid file in the GretinaGeometry with 1.5 mm back dead layer and
	3.5 mm coaxial dead layer from a minimization of the residual
	between simulated photopeak efficiencies (152Eu and 56Co) and
	those measured at the NSCL in September 2012.

Added an optional level scheme file specifying the portion of the
	level scheme of the reaction product accessible via de-excitation
	of the initial state. Multiple decay branches, individual level
	lifetimes, and individual gamma-ray angular distributions are
	supported. (Individual gamma-ray angular distributions are
	implemented but untested.)

Updated Reaction.cc: The Reaction process now specifies its ProcessType
	(Decay) and ProcessSubType (DecayExt) for compatibility with the
	PhysicsListHelper class.

Improved walls file in the GretinaGeometry tree. The hexagonal front and
	back caps still generate convexity warnings, but they are now all
	below the 0.1 mm level.

================================================================================
Version 1.2.4
================================================================================
L. Riley  March 2013

Positions of gamma-ray interaction points are now consolidated with an
	energy-weighted sum in EventAction.cc. Consolidation is not done
	by default (PositionRes = 0. mm).

Added a sourceFrame "co56_2012" for the 56Co source (25mm x 25m x .4mm Fe)

Added "eu152_peaks" (9 selected peaks, equal intensities) and "co56_peaks"
	(11 selected peaks, equal intensities) source types. These are
	useful for photopeak efficiencies.

Updated Physics Lists for compatibility with future Geant4
	releases. (Switched from SetModel to SetEmModel method.)

Resolved issues with Geant4.9.6 in CConvexPolyhedron.cc. (G4Exception now
	requires at least four arguments.)

Updated Physics Lists for compatibility with future Geant4
	releases. (Switched from SetModel to SetEmModel method.)
================================================================================
Version 1.2.3
================================================================================
L. Riley  February 2013

Modified EventAction and EventAction_Messenger: interaction points can be
	written in crystal coordinates as expected in Mode 2 data.
	The world coordinate system can also be first rotated by Pi/2
	about z to match the GRETINA coordinate system.
================================================================================
Version 1.2.2
================================================================================
L. Riley  February 2013

Modified cluster, walls, and euler geometry files in the
	./GretinaGeometry/NSCL tree to arrange crystals to match
	the NSCL numbering scheme

Modified EventAction.cc to write crystal ID numbers to match the NSCL
	numbering scheme
================================================================================
Version 1.2.1
================================================================================
L. Riley  January 2013

Consolidated energies and positions for each event into interaction points
	that can be resolved by GRETINA to better simulate Mode 2 data.

================================================================================
Version 1.2
================================================================================
L. Riley  August 2012

Migrated the GRETINA array and NSCL mounting shell from v1.1 to the
	UCSeGA/UCCAESAR	skeleton.

Implemented the NSCL target sled.

================================================================================
Version 1.1
================================================================================
L. Riley, B. Roberts, B. Sadler   June-July 2012

Implementation of the GRETINA mounting shell used at the NSCL and the
	cryostats/LN Dewars

Improved outer wall model

================================================================================
Version 1.0
================================================================================
L. Riley   July 2011

Initial implementation of LH target and NSCL target


src/Target.cc
#ifndef LHTARGET
#include "Target.hh"

Target::Target(G4LogicalVolume* experimentalHall_log,Materials* mat)
{
  materials=mat;
  expHall_log=experimentalHall_log;

  Target_side_x=50*mm;
  Target_side_y=50*mm;
  Target_thickness=0.1*mm;
  TargetMaterial = materials->FindMaterial("G4_Galactic");

  Pos = new G4ThreeVector(0.,0.,0.);
  Rot = G4RotationMatrix::IDENTITY;
  Rot.rotateZ(45.*deg);
  NoRot = G4RotationMatrix::IDENTITY;
  NStep=20;

  sourceFrame = "";

}

Target::~Target()
{ delete target_limits;}

//-----------------------------------------------------------------------------
G4VPhysicalVolume* Target::Construct()
{

  aTarget = new G4Box("target",Target_side_x/2.,Target_side_y/2.,Target_thickness/2.);

  Target_log = new G4LogicalVolume(aTarget,TargetMaterial,"target_log",0,0,0);
  target_limits= new G4UserLimits();
  target_limits->SetMaxAllowedStep(Target_thickness/NStep);
  Target_log->SetUserLimits(target_limits);

  Target_phys = new G4PVPlacement(G4Transform3D(Rot,*Pos),Target_log,"target",expHall_log,false,0);

  G4Colour red (1.0,0.0, 0.0); 
  G4VisAttributes* Vis_6 = new G4VisAttributes(red);
  if(TargetMaterial->GetName() == "G4_Galactic"){
    Vis_6->SetVisibility(false);
  } else {
    Vis_6->SetVisibility(true);
    Vis_6->SetForceSolid(true);
  }

  Target_log->SetVisAttributes(Vis_6);

  return Target_phys;
}

//-----------------------------------------------------------------------------
void Target::setX(G4double X)
{
   Target_side_x=X;
   aTarget->SetXHalfLength(Target_side_x/2.);
   G4cout<<"----> Target side X is set to "<<G4BestUnit(2.*aTarget->GetXHalfLength(),"Length")<<G4endl;
}
//-----------------------------------------------------------------------------
void Target::setY(G4double Y)
{
   Target_side_y=Y;
   aTarget->SetYHalfLength(Target_side_y/2.);
   G4cout<<"----> Target side Y is set to "<<G4BestUnit(2.*aTarget->GetYHalfLength(),"Length")<<G4endl;
}
//-----------------------------------------------------------------------------
void Target::setZ(G4double Z)
{
   Target_thickness=Z;
   aTarget->SetZHalfLength(Target_thickness/2.);
   target_limits->SetMaxAllowedStep(Target_thickness/NStep);
   Target_log->SetUserLimits(target_limits);
   G4cout<<"----> Target thickness is set to "<<G4BestUnit(2.*aTarget->GetZHalfLength(),"Length")<<G4endl;
}
//-----------------------------------------------------------------------------
void Target::setNStep(G4int n)
{
   NStep=n;
   target_limits->SetMaxAllowedStep(Target_thickness/NStep);
   Target_log->SetUserLimits(target_limits);
   G4cout<<"----> Number of simulation steps in the target is set to "<<NStep<<G4endl;
}

//-----------------------------------------------------------------------------
void Target::Report()
{
  G4cout<<"----> Target material set to  "<<Target_log->GetMaterial()->GetName()<< G4endl;   
  G4cout<<"----> Target density:         "<<Target_log->GetMaterial()->GetDensity()<< G4endl;   
  G4cout<<"----> Target side X is set to "<<G4BestUnit(2.*aTarget->GetXHalfLength(),"Length")<<G4endl;
  G4cout<<"----> Target side Y is set to "<<G4BestUnit(2.*aTarget->GetYHalfLength(),"Length")<<G4endl;  
  G4cout<<"----> Target thickness is set to "<<G4BestUnit(2.*aTarget->GetZHalfLength(),"Length")<<G4endl;
  G4cout<<"----> Number of simulation steps in the target is set to "<<NStep<<G4endl;
}
//---------------------------------------------------------------------
void Target::setMaterial(G4String materialName)
{
  // search the material by its name 
  TargetMaterial = materials->FindMaterial(materialName);  
  Target_log->SetMaterial(TargetMaterial);
  G4cout<<"----> Target material set to     "<<Target_log->GetMaterial()->GetName()<< G4endl;        

  G4Colour red (1.0,0.0, 0.0); 
  G4VisAttributes* Vis_6 = new G4VisAttributes(red);
  if(TargetMaterial->GetName() == "G4_Galactic"){
    Vis_6->SetVisibility(false);
  } else {
    Vis_6->SetVisibility(true);
    Vis_6->SetForceSolid(true);
  }

  Target_log->SetVisAttributes(Vis_6);
         
}
//-------------------------------------------------------------------
void Target::setTargetReactionDepth(G4double depth)
{
  //  G4cout<<"\n----> The depth is "<<G4BestUnit(depth,"Length")<< G4endl;;
  target_limits->SetUserMinRange(depth);
}
//-----------------------------------------------------------------------------
void Target::SetPositionZ(G4double d)
{
  Pos->setZ(d);
  G4ThreeVector Sep;
  Sep.setX(Pos->getX());
  Sep.setY(Pos->getY()); 
  Sep.setZ(Pos->getZ());
  Target_phys->SetTranslation(Sep);
  d=Target_phys->GetTranslation().getZ();
  G4cout<<"----> Target position is set to "<<G4BestUnit(d,"Length")<<G4endl;
}
//---------------------------------------------------------------------
void Target::ScaleDensity(G4double scale)
{
  // search the material by its name 
  G4String name=TargetMaterial->GetName();
  G4double Z=TargetMaterial->GetZ();
  G4double A=TargetMaterial->GetA();
  G4double density=TargetMaterial->GetDensity();
  density*=scale;
  TargetMaterial=new G4Material(name, Z,A,density);
  Target_log->SetMaterial(TargetMaterial);
  G4cout<<"----> Target material set to     "<<Target_log->GetMaterial()->GetName()<< G4endl;  
  G4cout<<"----> Target Z set to            "<<Target_log->GetMaterial()->GetZ()<< G4endl;  
  G4cout<<"----> Target mole mass set to       "<<Target_log->GetMaterial()->GetA()/g*mole<<" g/mole"<< G4endl;  
  G4cout<<"----> Target density set to         "<<Target_log->GetMaterial()->GetDensity()/g*cm3<<" g/cm3"<< G4endl;     
             
}
//---------------------------------------------------------------------
void Target::setSourceFrame(G4String sF)
{
  if(Target_phys)
    delete Target_phys;
  
  sourceFrame = sF;

  if(sourceFrame == "eu152_Z2707"){

    frameMaterial = materials->FindMaterial("Al");
    //    frameThickness = 2.9*mm; // Used prior to 3/2012
    frameThickness = 1.2*mm;       // Dirk Weisshaar 3/4/2012
    frameInnerRadius = 3.8*cm/2.0; // Source data sheet
    frameOuterRadius = 5.4*cm/2.0; // Source data sheet
    tapeMaterial = materials->FindMaterial("G4_POLYETHYLENE"); // Source data sheet
    tapeThickness = 0.012*cm;      // Source data sheet

    euFrame = new G4Tubs("euFrame",frameInnerRadius,frameOuterRadius,frameThickness/2.,0.,360.*deg);
    euFrame_log = new G4LogicalVolume(euFrame,frameMaterial,"euFrame_log",0,0,0);
    euFrame_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos),euFrame_log,"euFrame",expHall_log,false,0);

    euTape = new G4Tubs("euTape",0.,frameInnerRadius,tapeThickness/2.,0.,360.*deg);
    euTape_log = new G4LogicalVolume(euTape,tapeMaterial,"euTape_log",0,0,0);
    euTape_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos),euTape_log,"euTape",expHall_log,false,0);

  } else if(sourceFrame == "cs137_E2879"){

    frameMaterial = materials->FindMaterial("Al");
    frameThickness = 0.7*mm;
    frameInnerRadius = 2.54*13./16.*cm/2.0;
    frameOuterRadius = 2.54*cm/2.0;
    tapeMaterial = materials->FindMaterial("G4_POLYETHYLENE");
    tapeThickness = 0.016*cm;

    csFrame = new G4Tubs("csFrame",frameInnerRadius,frameOuterRadius,frameThickness/2.,0.,360.*deg);
    csFrame_log = new G4LogicalVolume(csFrame,frameMaterial,"csFrame_log",0,0,0);
    csFrame_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos),csFrame_log,"csFrame",expHall_log,false,0);

    G4ThreeVector ringPos(0.,0.,frameOuterRadius-frameInnerRadius+frameThickness/2.0);

    csRing = new G4Tubs("csRing",frameOuterRadius-frameThickness,frameOuterRadius,frameOuterRadius-frameInnerRadius,0.,360.*deg);
    csRing_log = new G4LogicalVolume(csRing,frameMaterial,"csRing_log",0,0,0);
    csRing_phys = new G4PVPlacement(G4Transform3D(NoRot,ringPos),csRing_log,"csRing",expHall_log,false,0);

    // Let's assume Mylar = Kapton = polyethylene is good enough.
    csTape = new G4Tubs("csTape",0.,frameInnerRadius,tapeThickness/2.,0.,360.*deg);
    csTape_log = new G4LogicalVolume(csTape,tapeMaterial,"csTape_log",0,0,0);
    csTape_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos),csTape_log,"csTape",expHall_log,false,0);

  } else if(sourceFrame == "co56_2012"){

    frameMaterial = materials->FindMaterial("Fe");
    frameThickness = 0.05*mm;
    frameSide_x = 25.0*mm;
    frameSide_y = 25.0*mm;

    coFrame = new G4Box("coFrame", frameSide_x/2., frameSide_y/2., frameThickness/2.);
    coFrame_log = new G4LogicalVolume(coFrame,frameMaterial,"coFrame_log",0,0,0);
    coFrame_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos),coFrame_log,"coFrame",expHall_log,false,0);

  }

  G4cout<<"----> Including source frame: "<<sourceFrame<< G4endl;                 
}
//-------------------------------------------------------------------
void Target::setSled()
{

  sledMaterial = materials->FindMaterial("G10");

  G4double tolerance = 0.1*mm;

  sledFrameThickness  = 0.02*2.54*cm;
  sledFrameRmin       = 2.5*2.54/2.0*cm;
  sledFrameRmax       = 3.0*2.54/2.0*cm;

  sledRunnerThickness = 0.031*2.54*cm;
  sledRunnerLength    = 4.5*2.54*cm;
  sledRunnerHeight    = 1.5*2.54*cm;
  sledRunnerRmin      = 1.3*2.54*cm;
  sledRunnerRmax      = 2.942*2.54*cm;

  sledBarThickness = 0.25*2.54*cm;
  sledBarLength    = 3.375*2.54*cm;
  sledBarLA        = 0.5*2.54*cm;
  sledBarHA        = 0.75*2.54*cm;
  sledBarLB        = sledBarLength - 2*sledBarLA;
  sledBarHBC        = 0.25*2.54*cm;
  sledBarLC        = 0.25*2.54*cm;

  sledFrame = new G4Tubs("sledFrame", sledFrameRmin, sledFrameRmax, sledFrameThickness/2.0, 0., 360.*deg);

  sledFrame_log = new G4LogicalVolume(sledFrame, sledMaterial, "sledFrame_log", 0, 0, 0);

  G4ThreeVector   *Pos0 = new G4ThreeVector(0., 0., 
					    Target_thickness/2.0 
					    + sledFrameThickness/2.0 
					    + tolerance);
  *Pos0 += *Pos;
  sledFrame_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos0), sledFrame_log, "sledFrame_phys", expHall_log, false, 0);

  sledRunnerBox  = new G4Box("sledRunnerBox", sledRunnerLength/2.0, sledRunnerHeight/2.0, sledRunnerThickness/2.0);
  sledRunnerTubs = new G4Tubs("sledRunnerTubs", sledRunnerRmin, sledRunnerRmax, sledRunnerThickness, 0., 360.*deg);
  Pos0->set(0., 1.5*2.54*cm, 0.);
  sledRunner     = new G4IntersectionSolid("sledRunner", sledRunnerBox, sledRunnerTubs, G4Transform3D(NoRot,*Pos0));

  sledRunner_log = new G4LogicalVolume(sledRunner, sledMaterial, "sledRunner_log", 0, 0, 0);

  Pos0->set(0., -1.5*2.54*cm, 
	    -sledBarLength/2.0 
	    - sledRunnerThickness/2.0 
	    + Target_thickness/2.0 
	    + sledFrameThickness/2.0);
  *Pos0 += *Pos;
  sledRunner1_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos0), sledRunner_log, "sledRunner1_phys", expHall_log, false, 0);

  Pos0->set(0., -1.5*2.54*cm, 
	    sledBarLength/2.0 
	    + sledRunnerThickness/2.0 
	    + Target_thickness/2.0 
	    + sledFrameThickness/2.0
	    + 2.*tolerance);
  *Pos0 += *Pos;
  sledRunner2_phys = new G4PVPlacement(G4Transform3D(NoRot,*Pos0), sledRunner_log, "sledRunner2_phys", expHall_log, false, 0);

  sledBarBoxA    = new G4Box("sledBarBoxA", sledBarThickness/2.0, sledBarHA/2.0,  sledBarLA/2.0);
  sledBarBoxB    = new G4Box("sledBarBoxB", sledBarThickness/2.0, sledBarHBC/2.0, sledBarLB/2.0);
  sledBarBoxC    = new G4Box("sledBarBoxC", sledBarThickness/2.0, sledBarHBC/2.0, sledBarLC/2.0);

  sledBarBoxA_log = new G4LogicalVolume(sledBarBoxA, sledMaterial, "sledBarBoxA_log", 0, 0, 0);
  sledBarBoxB_log = new G4LogicalVolume(sledBarBoxB, sledMaterial, "sledBarBoxB_log", 0, 0, 0);
  sledBarBoxC_log = new G4LogicalVolume(sledBarBoxC, sledMaterial, "sledBarBoxC_log", 0, 0, 0);

  sledBar        = new G4AssemblyVolume();
  
  sledBarZ1        = -sledBarLength/2.0 + 0.25*2.54*cm;
  sledBarY1        = 0.;
  Pos0 = new G4ThreeVector(0., sledBarY1, sledBarZ1-tolerance);
  sledBar->AddPlacedVolume(sledBarBoxA_log, *Pos0, &NoRot);

  sledBarZ2        = 0.;
  sledBarY2        = 0.;
  Pos0->set(0., sledBarY2, sledBarZ2);
  sledBar->AddPlacedVolume(sledBarBoxB_log, *Pos0, &NoRot);

  sledBarZ3        = -sledFrameThickness/2.0 - sledBarLC/2.0;
  sledBarY3        = 0.25*2.54*cm;
  Pos0->set(0., sledBarY3+tolerance, sledBarZ3-tolerance);
  sledBar->AddPlacedVolume(sledBarBoxC_log, *Pos0, &NoRot);

  sledBarZ4        = sledFrameThickness/2.0 + sledBarLC/2.0;
  sledBarY4        = sledBarY3;
  Pos0->set(0., sledBarY4+tolerance, sledBarZ4+tolerance);
  sledBar->AddPlacedVolume(sledBarBoxC_log, *Pos0, &NoRot);

  sledBarZ5        = sledBarLength/2.0 - 0.25*2.54*cm;
  sledBarY5        = 0.;
  Pos0->set(0., sledBarY5, sledBarZ5+tolerance);
  sledBar->AddPlacedVolume(sledBarBoxA_log, *Pos0, &NoRot);

  Pos0->set(-(1.011 + 1.287)/2.0*2.54*cm, -(1.011 + 1.287)/2.0*2.54*cm,
	    Target_thickness/2.0 
	    + sledFrameThickness/2.0 
	    + tolerance);
  *Pos0 += *Pos;
  G4RotationMatrix Rot0 = G4RotationMatrix::IDENTITY;  
  Rot0.rotateZ(-45.*deg);
  sledBar->MakeImprint(expHall_log, *Pos0, &Rot0);

  Pos0->set( (1.011 + 1.287)/2.0*2.54*cm, -(1.011 + 1.287)/2.0*2.54*cm,
	    Target_thickness/2.0 
	     + sledFrameThickness/2.0
	     + tolerance);
  *Pos0 += *Pos;
  Rot0 = G4RotationMatrix::IDENTITY;  
  Rot0.rotateZ(45.*deg);
  sledBar->MakeImprint(expHall_log, *Pos0, &Rot0);

  G4cout << "----> Including the target sled." << G4endl;
}
//-------------------------------------------------------------------
#endif
